File: 0046-permutations/0046-permutations.go
func permute(nums []int) [][]int {

	n:=len(nums)

	diary:=[]int{}
	res:=[][]int{}
	vis:=make(map[int]bool)

	var fun func(int)

	fun=func(idx int){

		// Last wla stage ka case lelo
		if idx==n{

			temp:=make([]int,len(diary))
			copy(temp,diary)

			res=append(res,temp)
			return
		}

		for i:=0;i<n;i++{

			if !vis[nums[i]]{

				// Agar visited hogya hai
				vis[nums[i]]=true
				diary=append(diary,nums[i])

				// Explore karlo
				fun(idx+1)

				// Reverse karlo
				diary=diary[:len(diary)-1]
				vis[nums[i]]=false
			}
		}
	}

	// Recursive Function call karlo
	fun(0)

	return res
}

File: 0046-permutations/0046-permutations.cpp
class Solution {
public:
void fun(vector<int>&nums,int n,int idx, vector<int>&diary,vector<vector<int>>&res,unordered_map<int,bool>&vis){

    // Last wla stage ka case lelo 
    if(idx==n){
        res.push_back(diary);
        return;
    }

    for(int i=0;i<n;i++){
        if(!vis[nums[i]]){

            vis[nums[i]]=true;              // Agar visited hogya hai 
            diary.push_back(nums[i]);

            // Explore karlo 
            fun(nums,n,idx+1,diary,res,vis);

            // Reverse karlo 
            diary.pop_back();
            vis[nums[i]]=false;
        }
    }
}

    vector<vector<int>> permute(vector<int>& nums) {

        int n=nums.size();
        vector<int>diary;            // Shuru mei to empty hi hona hai
        vector<vector<int>>res;      //Resultant 2D array bna lo 
        unordered_map<int,bool>vis; 

        // recursive Fuction call karlo 
        fun(nums,n,0,diary,res,vis);

        return res;                 // Final resultant wla array return karlo 
    }
};